unicode support for headers

Auto-reviewers: @NiharikaRay @matthewwardrop @earthmancash @danfrankj

It's not possible to have unicode chars in headers. It's important to have it for non-english posts.

stacktrace:
```bash
(knowledge-repo) tarsis@fury-road [/.../projetos/bigdata/bigdata-knowledge-repo]  (master)
$ knowledge_repo --repo posts/ add globoesporte/impacto_tr.kp/knowledge.md -p esportes --update
Traceback (most recent call last):
  File "/Users/tarsis/.virtualenvs/knowledge-repo/bin/knowledge_repo", line 255, in <module>
    kp = knowledge_repo.KnowledgePost.from_file(args.filename, src_paths=args.src)
  File "/Users/tarsis/.virtualenvs/knowledge-repo/lib/python2.7/site-packages/knowledge_repo/post.py", line 350, in from_file
    kp = KnowledgePostConverter.for_file(cls(), filename, format=format, postprocessors=postprocessors).from_file(filename, **opts)
  File "/Users/tarsis/.virtualenvs/knowledge-repo/lib/python2.7/site-packages/knowledge_repo/converter.py", line 53, in wrapped
    KnowledgePostProcessor._get_subclass_for(postprocessor).process(kp)
  File "/Users/tarsis/.virtualenvs/knowledge-repo/lib/python2.7/site-packages/knowledge_repo/postprocessors/format_checks.py", line 33, in process
    field, type(headers[field]), typ)
AssertionError: Value for field `title` is of type <type 'unicode'>, and needs to be of type <type 'str'>.
```

--- a/knowledge_repo/app/app.py
+++ b/knowledge_repo/app/app.py
@@ -46,7 +46,7 @@ class KnowledgeFlask(Flask):
 
         # Configure database
         db_uri = db_uri or self.config.get('SQLALCHEMY_DATABASE_URI') or 'sqlite:///:memory:'
-        logger.debug("Using database: {}".format(db_uri))
+        logger.debug(u"Using database: {}".format(db_uri))
         self.config['SQLALCHEMY_DATABASE_URI'] = db_uri
         self.config['SQLALCHEMY_ECHO'] = repo.config.debug
         self.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
@@ -173,7 +173,7 @@ class KnowledgeFlask(Flask):
             for key, value in new_values.items():
                 args[key] = value
 
-            return '{}?{}'.format(request.path, url_encode(args))
+            return u'{}?{}'.format(request.path, url_encode(args))
 
         @self.template_global()
         def pagination_pages(current_page, page_count, max_pages=5, extremes=True):

--- a/knowledge_repo/app/deploy/common.py
+++ b/knowledge_repo/app/deploy/common.py
@@ -23,7 +23,7 @@ class KnowledgeDeployer(with_metaclass(SubclassRegisteringABCMeta, object)):
                  port=7000,
                  workers=4,
                  timeout=60):
-        assert isinstance(knowledge_builder, (str, types.FunctionType)), "Unknown builder type {}".format(type(knowledge_builder))
+        assert isinstance(knowledge_builder, (str, types.FunctionType)), u"Unknown builder type {}".format(type(knowledge_builder))
         self.knowledge_builder = knowledge_builder
         self.host = host
         self.port = port
@@ -40,11 +40,11 @@ class KnowledgeDeployer(with_metaclass(SubclassRegisteringABCMeta, object)):
             out = []
             for nl, cell in zip(self.knowledge_builder.__code__.co_freevars, self.knowledge_builder.__closure__):
                 if isinstance(cell.cell_contents, str):
-                    out.append('{} = "{}"'.format(nl, cell.cell_contents.replace('"', '\\"')))
+                    out.append(u'{} = "{}"'.format(nl, cell.cell_contents.replace('"', '\\"')))
                 else:
-                    out.append('{} = {}'.format(nl, cell.cell_contents))
+                    out.append(u'{} = {}'.format(nl, cell.cell_contents))
             out.append(textwrap.dedent(inspect.getsource(self.knowledge_builder)))
-            return '\n'.join(out)
+            return u'\n'.join(out)
         return self.knowledge_builder
 
     @property
@@ -69,7 +69,7 @@ class KnowledgeDeployer(with_metaclass(SubclassRegisteringABCMeta, object)):
 
         out = []
         out.append('import sys')
-        out.append('sys.path.insert(0, "{}")'.format(kr_path))
+        out.append(u'sys.path.insert(0, "{}")'.format(kr_path))
         out.append('import knowledge_repo')
 
         out.append(self.builder_str)
@@ -77,7 +77,7 @@ class KnowledgeDeployer(with_metaclass(SubclassRegisteringABCMeta, object)):
             out.append('app = %s()' % self.knowledge_builder.__name__)
 
         with open(tmp_path, 'w') as f:
-            f.write('\n'.join(out))
+            f.write(u'\n'.join(out))
 
         return tmp_dir
 

--- a/knowledge_repo/app/deploy/gunicorn.py
+++ b/knowledge_repo/app/deploy/gunicorn.py
@@ -22,7 +22,7 @@ class GunicornDeployer(BaseApplication, KnowledgeDeployer):
 
     def load_config(self):
         options = {
-            'bind': '{}:{}'.format(self.host, self.port),
+            'bind': u'{}:{}'.format(self.host, self.port),
             'workers': self.workers,
             'timeout': self.timeout
         }

--- a/knowledge_repo/app/deploy/uwsgi.py
+++ b/knowledge_repo/app/deploy/uwsgi.py
@@ -12,7 +12,7 @@ class uWSGIDeployer(KnowledgeDeployer):
 
     _registry_keys = ['uwsgi']
 
-    COMMAND = "uwsgi --http {socket} --plugin python --wsgi-file {path} --callable app --master --processes {processes} --threads {threads} --uid --gid"
+    COMMAND = u"uwsgi --http {socket} --plugin python --wsgi-file {path} --callable app --master --processes {processes} --threads {threads} --uid --gid"
 
     def run(self):
         if not self.app.supports_threads:
@@ -21,7 +21,7 @@ class uWSGIDeployer(KnowledgeDeployer):
         tmp_dir = self.write_temp_files()
 
         config = {
-            'socket': '{}:{}'.format(self.host, self.port),
+            'socket': u'{}:{}'.format(self.host, self.port),
             'processes': self.workers,
             'threads': 2,
             'timeout': self.timeout,
@@ -29,8 +29,8 @@ class uWSGIDeployer(KnowledgeDeployer):
         }
 
         try:
-            cmd = "cd {};".format(tmp_dir) + self.COMMAND.format(**config)
-            logger.info("Starting server with command:  " + " ".join(cmd))
+            cmd = u"cd {};".format(tmp_dir) + self.COMMAND.format(**config)
+            logger.info("Starting server with command:  " + u" ".join(cmd))
             subprocess.check_output(cmd, shell=True)
         finally:
             shutil.rmtree(tmp_dir)

--- a/knowledge_repo/app/index.py
+++ b/knowledge_repo/app/index.py
@@ -2,6 +2,7 @@ from __future__ import absolute_import
 import logging
 import datetime
 import threading
+from builtins import str
 from flask import has_app_context
 
 from .proxies import db_session, current_repo, current_app
@@ -114,12 +115,12 @@ def _update_index(app, force=False, reindex=False):
         # If UUID has changed, check if we can find it elsewhere in the repository, and if so update index path
         if post.uuid and ((post.path not in kr_dir) or (post.uuid != kr_dir[post.path].uuid)):
             if post.uuid in kr_uuids:
-                logger.info('Updating location of post: {} -> {}'.format(post.path, kr_uuids[post.uuid].path))
+                logger.info(u'Updating location of post: {} -> {}'.format(post.path, kr_uuids[post.uuid].path))
                 post.path = kr_uuids[post.uuid].path
 
         # If path of post no longer in directory, mark as unpublished
         if post.path not in kr_dir:
-            logger.info('Recording unpublished status for post at {}'.format(post.path))
+            logger.info(u'Recording unpublished status for post at {}'.format(post.path))
             post.status = current_repo.PostStatus.UNPUBLISHED
             continue
 
@@ -131,17 +132,17 @@ def _update_index(app, force=False, reindex=False):
         # Update metadata of post if required
         if reindex or (kp.revision > post.revision or not post.is_published or kp.uuid != post.uuid):
             if kp.is_valid():
-                logger.info('Recording update to post at: {}'.format(kp.path))
+                logger.info(u'Recording update to post at: {}'.format(kp.path))
                 post.update_metadata_from_kp(kp)
             else:
-                logger.warning('Update to post at "{}" is corrupt.'.format(kp.path))
+                logger.warning(u'Update to post at "{}" is corrupt.'.format(kp.path))
 
     # Add the new posts that remain in kr_dir
     for kp_path, kp in kr_dir.items():
         if not kp.is_valid():
-            logger.warning('New post at "{}" is corrupt.'.format(kp.path))
+            logger.warning(u'New post at "{}" is corrupt.'.format(kp.path))
             continue
-        logger.info('creating new post from path {}'.format(kp_path))
+        logger.info(u'creating new post from path {}'.format(kp_path))
         post = Post()
         db_session.add(post)
         db_session.flush()  # (matthew) Fix groups logic so this is not necessary

--- a/knowledge_repo/app/models.py
+++ b/knowledge_repo/app/models.py
@@ -1,6 +1,7 @@
 import os
 import sys
 import traceback
+from builtins import str
 from future.utils import raise_with_traceback
 from flask import current_app, request, g
 from flask_sqlalchemy import SQLAlchemy
@@ -11,6 +12,7 @@ import datetime
 from werkzeug.local import LocalProxy
 from sqlalchemy import func, distinct, and_, select, UniqueConstraint
 import logging
+import six
 
 from knowledge_repo._version import __version__
 from knowledge_repo.repository import KnowledgeRepository
@@ -126,9 +128,9 @@ class ErrorLog(db.Model):
         filename = os.path.relpath(filename, os.path.join(os.path.dirname(__file__), '..'))
         return ErrorLog(
             function=function,
-            location='{}:{}'.format(filename, linenumber),
-            message='{}: {}'.format(e.__class__.__name__, "; ".join(str(a) for a in e.args)),
-            traceback="\n".join(traceback.format_tb(tb))
+            location=u'{}:{}'.format(filename, linenumber),
+            message=u'{}: {}'.format(e.__class__.__name__, u"; ".join(str(a) for a in e.args)),
+            traceback=u"\n".join(traceback.format_tb(tb))
         )
 
     @classmethod
@@ -308,7 +310,7 @@ class Tag(db.Model):
     def description(self):
         if self._description:
             return self._description
-        return "All posts with tag '{}'.".format(self.name)
+        return u"All posts with tag '{}'.".format(self.name)
 
     @description.expression
     def description(self):
@@ -374,7 +376,7 @@ class Post(db.Model):
 
     @hybrid_property
     def authors_string(self):
-        return ', '.join([author.format_name for author in self.authors])
+        return u', '.join([author.format_name for author in self.authors])
 
     @authors_string.expression
     def authors_string(self):

--- a/knowledge_repo/app/routes/editor.py
+++ b/knowledge_repo/app/routes/editor.py
@@ -126,7 +126,7 @@ def save_post():
 
     if prefixes is not None:
         if not any([path.startswith(prefix) for prefix in prefixes]):
-            return json.dumps({'msg': ("Your post path must begin with one of {}").format(prefixes),
+            return json.dumps({'msg': (u"Your post path must begin with one of {}").format(prefixes),
                                'success': False})
 
     # TODO better handling of overwriting
@@ -134,7 +134,7 @@ def save_post():
     if path in current_repo:
         kp = current_repo.post(path)
         if g.user.username not in kp.headers['authors'] and g.user.username not in current_repo.config.editors:
-            return json.dumps({'msg': ("Post with path {} already exists and you are not an author!"
+            return json.dumps({'msg': (u"Post with path {} already exists and you are not an author!"
                                        "\nPlease try a different path").format(path),
                                'success': False})
 
@@ -187,7 +187,7 @@ def publish_post():
     """ Publish the post by changing the status """
     path = request.args.get('path', None)
     if path not in current_repo:
-        return json.dumps({'msg': "Unable to retrieve post with path = {}!".format(path), 'success': False})
+        return json.dumps({'msg': u"Unable to retrieve post with path = {}!".format(path), 'success': False})
     current_repo.publish(path)
 
     update_index(check_timeouts=False)
@@ -200,7 +200,7 @@ def unpublish_post():
     """ Unpublish the post """
     path = request.args.get('path', None)
     if path not in current_repo:
-        return json.dumps({'msg': "Unable to retrieve post with path = {}!".format(path), 'success': False})
+        return json.dumps({'msg': u"Unable to retrieve post with path = {}!".format(path), 'success': False})
     current_repo.unpublish(path)
 
     update_index(check_timeouts=False)
@@ -213,7 +213,7 @@ def accept():
     """ Accept the post """
     path = request.args.get('path', None)
     if path not in current_repo:
-        return json.dumps({'msg': "Unable to retrieve post with path = {}!".format(path), 'success': False})
+        return json.dumps({'msg': u"Unable to retrieve post with path = {}!".format(path), 'success': False})
     current_repo.accept(path)
     update_index()
     return 'OK'
@@ -225,7 +225,7 @@ def delete_post():
     """ Delete a post """
     path = request.args.get('path', None)
     if path not in current_repo:
-        return json.dumps({'msg': "Unable to retrieve post with path = {}!".format(path), 'success': False})
+        return json.dumps({'msg': u"Unable to retrieve post with path = {}!".format(path), 'success': False})
     kp = current_repo.post(path)
     if g.user.username not in kp.headers['authors']:
         return json.dumps({'msg': "You can only delete a post where you are an author!", 'success': False})
@@ -290,7 +290,7 @@ def file_upload():
                     send_from_directory(dst_folder, filename)
                     uploadedFiles += [url_for("static", filename=os.path.join(upload_folder, filename))]
                 except Exception as e:
-                    error_msg = "ERROR during image upload: {}".format(str(e))
+                    error_msg = u"ERROR during image upload: {}".format(str(e))
                     logger.error(error_msg)
                     return json.dumps({'error_msg': error_msg, 'success': False})
 
@@ -302,11 +302,11 @@ def file_upload():
                     num_pages = src_pdf.getNumPages()
                     for page_num in range(num_pages):
                         page_png = pdf_page_to_png(src_pdf, page_num)
-                        page_name = "{filename}_{page_num}.jpg".format(**locals())
+                        page_name = u"{filename}_{page_num}.jpg".format(**locals())
                         page_png.save(filename=os.path.join(dst_folder, page_name))
                         uploadedFiles += [url_for("static", filename=os.path.join(upload_folder, page_name))]
                 except Exception as e:
-                    error_msg = "ERROR during pdf upload: {}".format(str(e))
+                    error_msg = u"ERROR during pdf upload: {}".format(str(e))
                     logger.error(error_msg)
                     return json.dumps({'error_msg': error_msg, 'success': False})
 

--- a/knowledge_repo/app/routes/index.py
+++ b/knowledge_repo/app/routes/index.py
@@ -9,6 +9,7 @@ This includes:
 import os
 import posixpath
 import json
+from builtins import str
 from flask import request, render_template, redirect, Blueprint, current_app, make_response
 from sqlalchemy import case, desc
 
@@ -39,7 +40,7 @@ def site_map():
         # url = url_for(rule.endpoint, **(rule.defaults or {}))
         links.append((str(rule), rule.endpoint))
     # links is now a list of url, endpoint tuples
-    return '<br />'.join(str(link) for link in links)
+    return u'<br />'.join(str(link) for link in links)
 
 
 @blueprint.route('/')
@@ -169,7 +170,7 @@ def render_cluster():
         tuples = [(k, v) for (k, v) in folder_to_posts.items()]
 
     else:
-        raise ValueError("Group by `{}` not understood.".format(group_by))
+        raise ValueError(u"Group by `{}` not understood.".format(group_by))
 
     if sort_by == 'alpha':
         grouped_data = sorted(tuples, key=lambda x: x[0])
@@ -247,5 +248,5 @@ def generate_users_typeahead():
 @blueprint.route('/ajax_paths_typeahead', methods=['GET'])
 def generate_projects_typeahead():
     # return path stubs for all repositories
-    stubs = ['/'.join(p.split('/')[:-1]) for p in current_repo.dir()]
+    stubs = [u'/'.join(p.split('/')[:-1]) for p in current_repo.dir()]
     return json.dumps(list(set(stubs)))

--- a/knowledge_repo/app/routes/posts.py
+++ b/knowledge_repo/app/routes/posts.py
@@ -1,5 +1,6 @@
 import logging
 import os
+from builtins import str
 from flask import request, url_for, redirect, render_template, current_app, Blueprint, g, Response
 
 from ..proxies import db_session, current_repo
@@ -51,7 +52,7 @@ def render(path):
                               .filter(Post.path == knowledge_aliases[path])
                               .first())
     if not post:
-        raise Exception("unable to find post at {}".format(path))
+        raise Exception(u"unable to find post at {}".format(path))
 
     if post.contains_excluded_tag:
         # It's possible that someone gets a direct link to a post that has an excluded tag
@@ -139,7 +140,7 @@ def _render_preview(path, tmpl):
             post = current_repo.post(knowledge_aliases[path])
 
     if not post:
-        raise Exception("unable to find post at {}".format(path))
+        raise Exception(u"unable to find post at {}".format(path))
 
     rendered = render_post(post, with_toc=True)
     raw_post = render_post_raw(post) if (mode == 'raw') else None
@@ -152,7 +153,7 @@ def _render_preview(path, tmpl):
                            raw_post=raw_post,
                            comments=[],
                            username=None,
-                           post_author=', '.join(post.headers['authors']),
+                           post_author=u', '.join(post.headers['authors']),
                            title=post.headers['title'],
                            page_views=0,
                            unique_views=0,
@@ -194,13 +195,13 @@ def download():
         return Response(
             post.to_string(format=resource_type),
             mimetype="application/zip",
-            headers={"Content-disposition": "attachment; filename={}".format(filename)})
+            headers={u"Content-disposition": "attachment; filename={}".format(filename)})
     elif resource_type == 'source':
         path = request.args.get('path', None)
         assert path is not None, "Source path not provided."
         return Response(
             post.read_src(path),
             mimetype="application/octet-stream",
-            headers={"Content-disposition": "attachment; filename={}".format(os.path.basename(path))})
+            headers={u"Content-disposition": "attachment; filename={}".format(os.path.basename(path))})
     else:
-        raise RuntimeError("Invalid resource_type: {}".format(resource_type))
+        raise RuntimeError(u"Invalid resource_type: {}".format(resource_type))

--- a/knowledge_repo/app/routes/tags.py
+++ b/knowledge_repo/app/routes/tags.py
@@ -15,6 +15,7 @@ from flask import current_app, request, render_template, Blueprint, g
 from sqlalchemy import and_
 import logging
 import math
+from builtins import str
 
 from ..app import db_session
 from ..models import PageView, Post, assoc_post_tag, Subscription, Tag

--- a/knowledge_repo/app/utils/emails.py
+++ b/knowledge_repo/app/utils/emails.py
@@ -88,7 +88,7 @@ def send_subscription_email(post, tag):
         return
 
     default_recipients = ['knowledge_consumer@notreal.com']
-    subject = "New knowledge post with tag [{}]!".format(tag.name)
+    subject = u"New knowledge post with tag [{}]!".format(tag.name)
     msg = Message(subject=subject,
                   recipients=default_recipients,
                   bcc=recipients_bcc)

--- a/knowledge_repo/app/utils/posts.py
+++ b/knowledge_repo/app/utils/posts.py
@@ -5,6 +5,7 @@ Functions include:
     - get_all_post_stats
 """
 import math
+from builtins import str
 from flask import current_app
 from sqlalchemy import func, distinct, or_
 

--- a/knowledge_repo/app/utils/render.py
+++ b/knowledge_repo/app/utils/render.py
@@ -45,8 +45,8 @@ def render_post_header(post):
     """
 
     def get_authors(usernames, authors):
-        authors = ["<a href='{}'>{}</a>".format(url_for('index.render_feed', authors=username), author) for username, author in zip(usernames, authors)]
-        return ' and '.join(', '.join(authors).rsplit(', ', 1))
+        authors = [u"<a href='{}'>{}</a>".format(url_for('index.render_feed', authors=username), author) for username, author in zip(usernames, authors)]
+        return u' and '.join(u', '.join(authors).rsplit(', ', 1))
 
     if isinstance(post, KnowledgePost):
         return header_template.format(title=post.headers['title'],

--- a/knowledge_repo/app/utils/search.py
+++ b/knowledge_repo/app/utils/search.py
@@ -30,5 +30,5 @@ def get_keywords(post):
     tldr = post.tldr.split(" ")
     tags = [tag.name for tag in post.tags]
     full_search_text = author + title + tldr + tags
-    keywords = ' '.join([word for word in full_search_text if word not in ENGLISH_STOPWORDS])
+    keywords = u' '.join([word for word in full_search_text if word not in ENGLISH_STOPWORDS])
     return keywords

--- a/knowledge_repo/config.py
+++ b/knowledge_repo/config.py
@@ -42,11 +42,11 @@ class KnowledgeRepositoryConfig(dict):
                     self.__update_from_file(value)
                 else:
                     logger.warning(
-                        "Configuration file {} does not exist.".format(value))
+                        u"Configuration file {} does not exist.".format(value))
             elif isinstance(value, type(None)):
                 pass
             else:
-                raise ValueError("Cannot interpret {}".format(value))
+                raise ValueError(u"Cannot interpret {}".format(value))
         dict.update(self, kwargs)
 
     def update_defaults(self, *values, **kwargs):
@@ -59,11 +59,11 @@ class KnowledgeRepositoryConfig(dict):
                 if os.path.exists(value):
                     self.__defaults_from_file(value)
                 else:
-                    logger.warning("Configuration file {} does not exist.".format(value))
+                    logger.warning(u"Configuration file {} does not exist.".format(value))
             elif isinstance(value, type(None)):
                 pass
             else:
-                raise ValueError("Cannot interpret {}".format(value))
+                raise ValueError(u"Cannot interpret {}".format(value))
         self.DEFAULT_CONFIGURATION.update(kwargs)
 
     def __defaults_from_file(self, filename):
@@ -79,7 +79,7 @@ class KnowledgeRepositoryConfig(dict):
         self.__set_from_module(self, module)
 
     def __set_from_file(self, d, filename, force=False):
-        config = imp.load_source('knowledge_repo.config_{}'.format(str(time.time()).replace('.', '')), filename)
+        config = imp.load_source(u'knowledge_repo.config_{}'.format(str(time.time()).replace('.', '')), filename)
         self.__set_from_module(d, config, force)
 
     def __set_from_module(self, d, module, force=False):

--- a/knowledge_repo/config_defaults.py
+++ b/knowledge_repo/config_defaults.py
@@ -42,7 +42,7 @@ def username_to_name(username):
 
 # Function to convert a username to a person's email
 def username_to_email(username):
-    return '{}@example.com'.format(username)
+    return u'{}@example.com'.format(username)
 
 
 # Function to generate the web uri for a knowledge post at

--- a/knowledge_repo/converters/html.py
+++ b/knowledge_repo/converters/html.py
@@ -80,7 +80,7 @@ class IndentsAsCellOutputProcessor(BlockProcessor):
 
             block_is_html = block_is_html and not isinstance(sibling.text, AtomicString)
 
-            block = '\n'.join([sibling.text, block])
+            block = u'\n'.join([sibling.text, block])
             output = sibling
         else:
             # This is a new codeblock. Create the elements and insert text.

--- a/knowledge_repo/mapping.py
+++ b/knowledge_repo/mapping.py
@@ -29,7 +29,7 @@ class SubstitutionMapper(object):
             else:
                 output.append(replacement)
         output.append(text[last_offset:])
-        return ''.join(output)
+        return u''.join(output)
 
     def find_matches(self, text, reverse=False):
         matches = []

--- a/knowledge_repo/postprocessors/format_checks.py
+++ b/knowledge_repo/postprocessors/format_checks.py
@@ -1,22 +1,24 @@
 import datetime
+
+import six
 from past.builtins import basestring
 
 from ..postprocessor import KnowledgePostProcessor
 
 REQUIRED_FIELD_TYPES = {
-    'title': str,
+    'title': six.string_types,
     'authors': list,
     'created_at': datetime.datetime,
-    'tldr': basestring,
+    'tldr': six.string_types,
     'tags': list
 }
 
 OPTIONAL_FIELD_TYPES = {
-    'path': str,
+    'path': six.string_types,
     'updated_at': datetime.datetime,
     'private': bool,   # If true, this post starts out private
     'allowed_groups': list,
-    'thumbnail': (int, str)
+    'thumbnail': (int, ) + six.string_types
 }
 
 

--- a/knowledge_repo/repositories/dbrepository.py
+++ b/knowledge_repo/repositories/dbrepository.py
@@ -32,7 +32,7 @@ class DbKnowledgeRepository(KnowledgeRepository):
 
         # TODO handle if user does not pass in table sqlite://path.db
         uri_splt = self.uri.split(":")
-        engine_uri = ":".join(uri_splt[:-1])
+        engine_uri = u":".join(uri_splt[:-1])
         table_name = uri_splt[-1]
 
         metadata = MetaData()

--- a/knowledge_repo/repositories/gitrepository.py
+++ b/knowledge_repo/repositories/gitrepository.py
@@ -203,7 +203,7 @@ class GitKnowledgeRepository(KnowledgeRepository):
                 break
         if not ref.endswith('.kp'):
             return None
-        return '/'.join(refs[:i + 1])
+        return u'/'.join(refs[:i + 1])
 
     def git_local_posts(self, branches=None, as_dict=False):
         if branches is None:

--- a/knowledge_repo/utils/dependencies.py
+++ b/knowledge_repo/utils/dependencies.py
@@ -14,4 +14,4 @@ def check_dependencies(dependencies, message=None):
                "\t{install_command}\n\n"
                "Note: Depending on your system's installation of Python, you may "
                "need to use `pip2` or `pip3` instead of `pip`.").format(install_command='pip install --upgrade ' + ' '.join(missing_deps))
-        raise RuntimeError('\n\n'.join([message, fix]))
+        raise RuntimeError(u'\n\n'.join([message, fix]))

--- a/run_tests.sh
+++ b/run_tests.sh
@@ -45,6 +45,12 @@ for post in $(ls `dirname $0`/tests/test_posts); do
     if [[ "${post}" == *.ipynb ]]; then
         `dirname $0`/scripts/knowledge_repo --repo="${test_repo_path}" --dev add `dirname $0`/tests/test_posts/${post} -p projects/${post} -m "Test commit" --branch master;
     fi;
+    if [[ "${post}" == *.Rmd ]]; then
+        `dirname $0`/scripts/knowledge_repo --repo="${test_repo_path}" --dev add `dirname $0`/tests/test_posts/${post} -p projects/${post} -m "Test commit" --branch master;
+    fi;
+    if [[ "${post}" == *.md ]]; then
+        `dirname $0`/scripts/knowledge_repo --repo="${test_repo_path}" --dev add `dirname $0`/tests/test_posts/${post} -p projects/${post} -m "Test commit" --branch master;
+    fi;
 done
 
 echo

--- a/tests/test_posts/unicode.md
+++ b/tests/test_posts/unicode.md
@@ -1,0 +1,15 @@
+---
+title: 我能说普通话
+authors:
+    - 吴文浩
+created_at: 2017-07-03
+updated_at: 2017-07-03
+tags:
+    - 标签
+tldr: |
+    书到用时方恨少
+---
+
+## 我们支持unicode
+
+欢呼！

